etl/steps/data/garden/energy/2022-09-09/global_primary_energy.py

```diff
@@ -0,0 +1,223 @@
+# All other energy sources in Smil's data will not be affected by this factor.
+EFFICIENCY_FACTOR = 0.36
+
+
+def prepare_bp_data(tb_bp: catalog.Table) -> pd.DataFrame:
+    df_bp = pd.DataFrame(tb_bp).reset_index()
+
+    # BP gives generation of direct energy in TWh, and, for non-fossil sources of electricity,
+    # consumption of input-equivalent energy in EJ.
+    # The input-equivalent energy is the amount of energy that would be required to generate a given amount of (direct)
+    # electricity if non-fossil sources were as inefficient as a standard thermal power plant.
+    # Therefore, direct and substituted energies for Biofuels, Coal, Gas and Oil are identical.
+    # On the other hand, direct and substituted energy are different for non-fossil electricity sources, namely
+    # Hydropower, Nuclear, Solar, Other renewables, and Wind.
+    # The difference is of a factor of ~38%, which is roughly the efficiency of a standard power plant.
+    # More specifically, BP assumes (for Biofuels, Coal, Gas and Oil) an efficiency factor that grows from 36%
+    # (until year 2000) to 40.6% (in 2021), to better reflect changes in efficiency over time.
+    # In the case of biomass used in electricity (included in 'Other renewables'),
+    # BP assumes a constant factor of 32% for all years.
+    # For more details:
+    # https://www.bp.com/content/dam/bp/business-sites/en/global/corporate/pdfs/energy-economics/statistical-review/bp-stats-review-2022-methodology.pdf
+    bp_columns = {
+        "country": "country",
+        "year": "year",
+        # Fossil sources (direct energy).
+        "biofuels_consumption__twh__total": "biofuels__twh_direct_energy",
+        "coal_consumption__twh": "coal__twh_direct_energy",
+        "gas_consumption__twh": "gas__twh_direct_energy",
+        "oil_consumption__twh": "oil__twh_direct_energy",
+        # Non-fossil electricity sources (direct energy).
+        "geo_biomass_other__twh": "other_renewables__twh_direct_energy",
+        "hydro_generation__twh": "hydropower__twh_direct_energy",
+        "nuclear_generation__twh": "nuclear__twh_direct_energy",
+        "solar_generation__twh": "solar__twh_direct_energy",
+        "wind_generation__twh": "wind__twh_direct_energy",
+        # Non-fossil electricity sources (substituted energy).
+        "geo_biomass_other__ej": "other_renewables__ej_substituted_energy",
+        "hydro_consumption__ej": "hydropower__ej_substituted_energy",
+        "nuclear_consumption__ej": "nuclear__ej_substituted_energy",
+        "solar_consumption__ej": "solar__ej_substituted_energy",
+        "wind_consumption__ej": "wind__ej_substituted_energy",
+    }
+    df_bp = df_bp[list(bp_columns)].rename(columns=bp_columns)
+    # Convert all units to TWh.
+    for column in df_bp.columns:
+        if "_ej_" in column:
+            # Create a new column in TWh instead of EJ.
+            df_bp[column.replace("_ej_", "_twh_")] = df_bp[column] * EJ_TO_TWH
```

Comment:
You could have `df_bp.pop(column)  * EJ_TO_TWH` to get rid of the next two lines, but whatever.

Reply:
I don't like this solution very much.